Fixed xDscResourceDesigner blank exit code installation failures. Fixes #8

PlagueHO opened the pull request:
This should fix Issue #8.

The following changes were made:
1. Code to download xDSCResourceDesigner moved from Meta.Tests.ps1 into Get-ResourceDesigner function of TestHelper.psm1
2. If PowerShellGet module installed: xDSCResourceDesigner downloaded using Install-Module.
3. If PowerShellGet Module not installed: xDSCResourceDesigner install using Nuget.exe.
4. No longer uses Start-Process to call out to Nuget.exe.
5. xDSCResourceDesigner only downloaded automatically when running in AppVeyor.
6. xDSCResourceDesigner Module imported using Name only (not path) so that it works when installed via PowerShellGet.
7. Added some AppVeyor config samples and other details into the Readme.md.

I've thrown this through a lot of tests on projects using both Windows Server 2012 R2 and Unstable AppVeyor build environments and have not run into any failures so far.

Thanks!

msftclas commented:
Hi __@PlagueHO__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -73,3 +73,61 @@ function New-Nuspec
+    }
+
+    # Import the module using the name
+    Import-Module -Name $DesignerModuleName -Force
```
Does that mean if we install tests on non-appveyor machine, we need to install xDscResourceDesigner manually before running it or tests won't run at all?

PlagueHO replied:
It does, yes:
![2015-11-06_10-15-48](https://cloud.githubusercontent.com/assets/7589164/10981962/8efdf9ee-846f-11e5-838b-82326bbdda98.png)
I kept it simple as suggested so that errors like that are just reported straight back to the user.

It might be nicer to detect the missing module and report it to the user as a Write-Warning so they know what is wrong. But my thinking is it is still better that the tests fail so that CI fails (assuming they're not using AppVeyor).

KarolKaczmarek replied:
I'm not convinced this is the right approach. Think about it this way:

A new contributor wants to contribute to xSharePoint (or any other resource), clones the repo, runs tests and half of them fail. There's no information why they fail other than "Test-xDscResource" is not recognized. They likely are not aware of what this cmdlet does and that it belongs to the xDscResourceDesigner...

I'd find this experience very discouraging. 

At the very least, we should detect whether Import-Module fails and if it does (and we aren't on appveyor), display warning that we may need to install xDscResourceDesigner manually.

Having said that, I'd actually say that there's nothing wrong in downloading xDscResourceDeisgner on non-appveyor machines. After all, our tests are modifying states of users' machines and they should be aware of that when running them (or any other script for that matter).

That'd be my preference, but I'd like to hear some other opinions on this before proceeding. @TravisEz13 ?

PlagueHO replied:
Because a lot of resources I work on need to be tested locally on specific servers (because of Windows Feature requirements) which are often clean and don't have the module, I'd love for the automatic installation of the module if it is missing to be the default. The **shouldprocess** solution seems like it might be a nice solution to this.

Also, if the user rejects installing it they should at least be presented with a legible warning that the tests will fail so they don't get discouraged outright.

KarolKaczmarek replied:
Great, I think you can proceed with this implementation. 
If user rejects the confirmation and sees the warning, that should give him pretty good hint of what may be causing the failures ;)

When testing your changes, could you make sure the -force option allows to skip the confirmation prompt so that we can use it in a fully automated way to test all Dsc Resources?

KarolKaczmarek commented on Meta.Tests.ps1:
```diff
@@ -60,6 +28,9 @@ if (($env:PSModulePath.Split(';') | Select-Object -First 1) -ne $pwd)
 }
 
+# Make sure MetaFixers.psm1 is loaded - it contains Get-TextFilesList
+Import-Module -Name (Join-Path -Path $PSScriptRoot -ChildPath 'MetaFixers.psm1') -Force
```
It's good practice to leave all Import-Module at the top of the script to easily find out what dependencies script uses. Can you please move it back?

PlagueHO replied:
Good to know! Done.

PlagueHO commented:
Regarding automatically installing the ResourceDesigner in a **Non-AppVeyor** environment and WMF5.0 is **not** installed means that Nuget.exe may not be available. If it is not available then it could be downloaded directly from http://nuget.org/nuget.exe using Invoke-WebRequest (ShouldProcess here of course) - that way WMF3.0 and WMF4.0 users can still use automatic installation of the module etc. Of course WMF2.0 users are out of luck here (no Invoke-Webrequest) and an error would be thrown, but does any of this work on WMF2.0 really?

Is this acceptable behavior?

PlagueHO commented:
I've made the suggested changes to the PR.

However, one issue regarding what @KarolKaczmarek suggested re the Force switch:

I've implemented the switch no problem, but the problem is that Meta.Tests.ps1 is actually called by Invoke-Pester, so as far as I know we can't pass parameters to the actual tests scripts (or is there a way to do this). So actually setting the -force parameter isn't possible. So this means the user will be asked to confirm downloading of Nuget and xDSCResourceDesigner.

There are two ways around this:
1. Set ```$ConfirmPreference='None'``` before calling the pester tests - but this makes me cringe and I'd recommend against it.
2. Alternately, because this code has been moved into the TestHelper.psm1, the following just needs to be fired once before calling Invoke-Pester:
```
Import-Module TestHelper.psm1
Import-ResourceDesigner -force
```
So this will allow both CI scenarios (non-appveyor) and user interactive scenarios, so the user won't be just thrown a bunch of errors and give up.

This will install the module silently (no confirmation) if it is not installed and then import it. If it is already installed it'll just import it.

KarolKaczmarek commented:
I think we can go with the second option, I assume there's a typo and you meant Install-ResourceDesigner -force, not Import-ResourceDesigner -force?

PlagueHO commented:
Totally agree - option 1 should be not mentioned really. And that is correct, I meant ```Install-ResourceDesigner``` - sorry about that. I've got rid of the Import-ResourceDesigner function now to simplify things.

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -73,3 +79,159 @@ function New-Nuspec
+                -f $Script:DesignerModuleName,$nugetSource)))
+        {
+            # Use Nuget.exe to install the module
+            & "$nugetPath" @( `
```
Does this method have 100% success rate after replacing Start-Process with "&"?
If so, is there a reason why we verify whether PackageManagement/PowerShellGet are available and use them when that's the case? 
Couldn't we just reliably use nugget.exe in all cases which would simplify the logic here?

PlagueHO replied:
It does have a 100% success rate in all my tests (both in local testing and AppVeyor). So yes, we could definitely do away with the PowerShellGet method and always use Nuget.

I think there was some suggestion and some point to support both methods, but we can definitely keep it simple and get rid of that code path.

Consider it done!

KarolKaczmarek replied:
Thanks!

KarolKaczmarek commented on README.md:
````diff
@@ -54,3 +54,30 @@ A fixer corresponds to a particular test.
+To automatically download and install the DscResource.Tests in an AppVeyor.yml file, add
+the following to the install section:
+
+```
````
I'd recommend linking to the appveyor.yml example here instead of copying the snippets to get rid of redunancy: https://github.com/PowerShell/DscResources/blob/master/DscResource.Template/appveyor.yml 

It contains the same code, but with broader context, so let's keep it in one place.

PlagueHO replied:
:+1:

KarolKaczmarek commented on Meta.Tests.ps1:
```diff
@@ -1,57 +1,52 @@
+# Install and/or Import xDSCResourceDesigner Module
+if ($env:APPVEYOR) {
+    # Running in AppVeyor so force silent install of xDSCResourceDesigner
+    $PSBoundParameters.Force = $true
```
Looks like because of this the workaround which you mentioned in the conversation is no longer necessary, right?

```
2. Alternately, because this code has been moved into the TestHelper.psm1, the following just needs to be fired once before calling Invoke-Pester:
Import-Module TestHelper.psm1
Install-ResourceDesigner -force


So this will allow both CI scenarios (non-appveyor) and user interactive scenarios, so the user won't be just thrown a bunch of errors and give up.

This will install the module silently (no confirmation) if it is not installed and then import it. If it is already installed it'll just import it.

```

PlagueHO replied:
The check for $env:AppVeyor isn't strictly required, but if that code is removed all existing AppVeyor.yml configs would need to have the following line added:
```
Install-ResourceDesigner -force
```
Added. E.g.:
```
#---------------------------------#
#      environment configuration  #
#---------------------------------#
os: Unstable
version: 2.4.{build}.0
install:
  - cinst -y pester
  - git clone https://github.com/PowerShell/DscResource.Tests
  - ps: Push-Location
  - ps: Import-Module DscResource.Tests\TestHelper.psm1 -force
  - ps: Install-ResourceDesigner -force
.......
```

Which I think would be a real headache (and any ones that didn't get update will start failing). This code ensures that no matter what happens existing projects will continue to work as normal.

KarolKaczmarek replied:
That's great!

KarolKaczmarek commented:
Thank you! Code looks great. Let me do some more testing and I'll merge it soon.

PlagueHO commented:
Great! Thanks heaps for looking at this for me. I fixed one last bug that was exposed when I removed the PowerShellGet code path where Nuget.exe was putting text into the default stream.
